tianshou/data/her_collector.py

```diff
@@ -0,0 +1,304 @@
+
+                # remove surplus env id from ready_env_ids
+                # to avoid bias in selecting environments
+                if n_episode:
+                    surplus_env_num = len(ready_env_ids) - (n_episode - episode_count)
+                    if surplus_env_num > 0:
+                        mask = np.ones_like(ready_env_ids, dtype=bool)
+                        mask[env_ind_local[:surplus_env_num]] = False
+                        ready_env_ids = ready_env_ids[mask]
+                        self.data = self.data[mask]
+
+                # use HER to create more trajectory
+                for env_id in env_ind_global:  # enumerate env
+                    # get recently collected data from buffer
+                    env_buffer = self.buffer.buffers[env_id]
+                    env_buffer_len = env_buffer.last_index[0] + 1
+                    traj_len = ep_len[env_id]
+                    obs_index_range = np.arange(
+                        env_buffer_len - traj_len, env_buffer_len
+                    ) % len(env_buffer)
+                    original_trajectory = env_buffer[obs_index_range]
+                    if self.strategy == 'offline':
+                        new_trajactory_len = (
+                            np.random.random(size=self.replay_k) * traj_len
+                        ).astype(int) + 1
+                        # relabel data and add back
+                        for length in new_trajactory_len:
+                            trajectory = Batch(original_trajectory[:length], copy=True)
+                            new_goal = trajectory.obs_next[
+                                length - 1, self.obs_index_range['achieved_goal']]
+                            new_goals = np.repeat([new_goal], length, axis=0)
+                            trajectory.obs[:, self.
+                                           obs_index_range['desired_goal']] = new_goals
+                            trajectory.obs_next[:, self.obs_index_range['desired_goal']
+                                                ] = new_goals
+                            trajectory.rew = self.reward_fn(
+                                trajectory.obs_next[:, self.
+                                                    obs_index_range['achieved_goal']],
+                                new_goals, None
+                            )
+                            trajectory.done[-1] = True
+                            for i in range(length):
+                                env_buffer.add(trajectory[i])
+                    elif self.strategy == 'online':
+                        # record the achieved goal of future steps,
+                        # to reduce the relabel time during the trainning
+                        ag = original_trajectory.obs_next[:, self.obs_index_range[
+                            'achieved_goal']]
+                        for i, idx in enumerate(obs_index_range):
+                            env_buffer.info.achieved_goal[idx] = ag[i:]
```

Comment:
I don't think it is a good idea to implement HER on the collector side, instead, buffer would be the best fit. In my previous thinking, we can only call a function like `buffer.perform_her(...)` to solve this problem and it is more general and suitable to all algorithms (instead of creating something like SACHERPolicy).

Reply:
Yes, I agree with that. Relabeling in the buffer is the most intuitive way. There are two ways of implementing HER in other repositories. Both [stable baseline3](https://github.com/DLR-RM/stable-baselines3) and [Tianhong Dai](https://github.com/TianhongDai/hindsight-experience-replay) use online sampling, which means relabeling when sampling the data. This way can generate more fresh data than the original version of HER proposed by the paper, which relabels data in the buffer then sample it as usual. This led to two different ways of implementing `perform_her()`: relabel all data in the buffer or relabel newly collected data. 
Here are my a few concerns about When to perform `buffer.perform_her(...)` and How to perform `buffer.perform_her(...)`:
1. For an offline version of HER:
* Why I choose to modify `collector`: The current version is as follows: After the collection of one trajectory is over, I repeat it for k times, relabel then store it. I have not figured out how to identify the newly collected data and deal with the partially collected trajectory. (add newly relabeled data will break them). 
* Difficulties in implementing it in `buffer`: I think we can perform the relabel per collect by calling `perform_her()`. But I have not figured out how to identify the newly collected data. Moreover, suppose we only relabel the collected data and do not generate more trajectory. In that case, it will hurt the sample efficiency as lots of original data is lost, and more goal choice is ignored.
2. For the online version of HER:
* Why I choose to modify `policy`: My initial idea was to do it in `buffer.sample()` to get relabeled data. Still, I found that the `process_fn()` function only uses the subscript obtained by sampling rather than the data, so it is hard to relabel in the buffer's sample. That's why I had the idea to modify the policy `pre_process`. So I changed the policy `pre_process` function, which can be applied to any off-policy algorithm. I have to admit that it is not a clean way to do this, but it could be general.
* Difficulties to implement it in `buffer`: The `buffer` and `policy.process_fn()` is not disentangled. I understand this design is to achieve an n-step return, but I have not figured out a better solution for the online sampling in the buffer as the sampled data is not used.

I wonder if it is possible to store alternative goals in the `info` though `buffer.perform_her()`  and sample it when updating the network. (This is what I am doing in the online HER). It is efficient(saving a lot of time to sample a new goal in the buffer during the update) and can extend to other relabel methods (e.g., inverse RL). The main obstacle is we still need to redefine the `policy.process_fn()` to recompute reward and get a new goal, as it will get data directly from the buffer. (Or to modify the `buffer.sample()` to get n_step data and decouple the policy and buffer, use the sample function as the only portal to get data). 
Anyway, tianshou shows excellent potential to perform such training task quickly, especially in relabel algorithms like HER, as most repositories use MPI to parallelize, which is resource-consuming. Hindsight Experience Replay is a crucial method to solve sparse reward tasks, and it should not be missing in such a great training framework.
Any information will be highly appreciated.

Reply:
@jc-bao Is the online version almost always better than offline version? If so, I'd suggest we implement the online version only because I think modifying `collector` is much worse than modifying `policy`.

How about inheriting `buffer` and overriding `sample()` and/or related methods to perform relabeling on-the-fly?

If there is NO way to implement it as `buffer.perform_her()`, I think overriding `policy.process_fn()` is an acceptable compromise.

Reply:
1. Is the online version almost always better than the offline version?
Ans: Yes, the online version has 5 times higher sample efficiency compared to the original HER.
2. How about inheriting buffer and overriding sample()?
Ans: current version of tianshou will only use the index return from the sample function to sample data. The relabeled data will not be used. So it is insufficient to only modify the sample().
3. implement it as buffer.perform_her()
Ans: It is feasible for the offline version.
4. override process_fn():
Ans: [here](https://github.com/jc-bao/tianshou/blob/68f22cafc5e906d783d32964fa7f1aa91e69cf91/tianshou/policy/modelfree/sac_her.py#L92) is my solution. The problem is that it is not a general solution.

Reply:
> 4\. Ans: [here](https://github.com/jc-bao/tianshou/blob/68f22cafc5e906d783d32964fa7f1aa91e69cf91/tianshou/policy/modelfree/sac_her.py#L92) is my solution. The problem is that it is not a general solution.

What do you mean it's not general? If you mean you want to make HER available to any off policy methods, not just SAC, you can try implementing it as a [decorator](https://en.wikipedia.org/wiki/Decorator_pattern). See an example [here](https://github.com/thu-ml/tianshou/blob/master/tianshou/policy/modelbased/icm.py) and [how it is used](https://github.com/thu-ml/tianshou/blob/master/examples/atari/atari_ppo.py#L161-L167).

Reply:
Thanks for the information! I will try it later.